### PR TITLE
[Model] Support Cohere2ForCausalLM (Cohere R7B)

### PR DESCRIPTION
**Pull request by janimo**
Sliding window support as done in HuggingFace Transformers (>= 4.48.0.dev0) for the Cohere2ForCausalLM arch.

### CODE REVIEW

File: docs/source/models/supported_models.rst [hunk 1]
```diff
@@ -120,7 +120,7 @@ Text Generation (``--task generate``)
     - ✅︎
   * - :code:`CohereForCausalLM`
```
Comment (DarkLight1337) [inline docs/source/models/supported_models.rst:121]:
```suggestion
  * - :code:`CohereForCausalLM`, :code:`Cohere2ForCausalLM`
```

File: vllm/model_executor/models/commandr.py [hunk 2]
```diff
@@ -206,7 +215,8 @@ def forward(
         q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
         if self.use_qk_norm:
             q, k = self._apply_qk_norm(q, k)
-        q, k = self.rotary_emb(positions, q, k)
+        if self.sliding_window:
+            q, k = self.rotary_emb(positions, q, k)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/commandr.py:219]: I think you inverted the condition here? Since Cohere (v1) doesn't have sliding window.

Comment (janimo) [reply vllm/model_executor/models/commandr.py:219]: You're right, I'll fix this. The condition is ok for v2 but should not be there at all for v1.

File: docs/source/models/supported_models.rst (hunk 1 above)
Comment (janimo) [reply docs/source/models/supported_models.rst:121]: I thought the code meant the vllm classname (of which I did not make a new one for v2). If it is the architecture_name from the config then it makes sense.

Comment (DarkLight1337) [reply docs/source/models/supported_models.rst:121]: It should be the architecture name on HF side (so people can check the `config.json` and search for it in this list)

File: vllm/model_executor/models/commandr.py (hunk 2 above)
Comment (janimo) [reply vllm/model_executor/models/commandr.py]: @DarkLight1337 I have added the extra check.

**Comment by DarkLight1337** [comment]
Could you point to the place where rotary embedding is disabled in v2? Looking at the HF implementation, I don't see why it should be disabled when sliding window is active.

**Comment by DarkLight1337** [comment]
Never mind, I confused

https://github.com/huggingface/transformers/blob/main/src/transformers/models/cohere2/modeling_cohere2.py#L387

with 

https://github.com/huggingface/transformers/blob/main/src/transformers/models/cohere2/modeling_cohere2.py#L734

File: vllm/model_executor/models/commandr.py [hunk 3]
```diff
@@ -171,12 +171,26 @@ def __init__(
...
+        # Model v2 has sliding windows, v1 does not
+        self.v1 = self.sliding_window is None
+
+        layer_idx = extract_layer_index(prefix)
+        layer_has_sliding_window = (
+            getattr(config, "sliding_window_pattern", False)
+            and (layer_idx + 1) % self.config.sliding_window_pattern != 0)
+
+        if not layer_has_sliding_window:
+            self.sliding_window = None
```
Comment (DarkLight1337) [inline vllm/model_executor/models/commandr.py:185]:
```suggestion
        sliding_window = getattr(config, "sliding_window", None)
        # Model v2 has sliding windows, v1 does not
        self.v1 = sliding_window is None

        layer_idx = extract_layer_index(prefix)
        layer_has_sliding_window = (
            getattr(config, "sliding_window_pattern", False)
            and (layer_idx + 1) % self.config.sliding_window_pattern != 0)

        if layer_has_sliding_window:
            self.sliding_window = sliding_window
```

Let's avoid assigning to `self.sliding_window` prematurely.

Comment (janimo) [reply vllm/model_executor/models/commandr.py]: Done.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for adding support for this!

**Comment by DarkLight1337** [comment]
Please resolve the failing test.

**Comment by janimo** [comment]
I've added the model to the test list, but wonder if it's a problem if this arch is only in the unreleased version of transformers.

**Comment by DarkLight1337** [comment]
Can you update `tests/models/test_initialization.py::test_can_initialize` to skip this model if the transformers version is too old?

File: tests/models/test_initialization.py [hunk 4]
```diff
@@ -11,6 +12,9 @@
 @pytest.mark.parametrize("model_arch", HF_EXAMPLE_MODELS.get_supported_archs())
 def test_can_initialize(model_arch):
     model_info = HF_EXAMPLE_MODELS.get_hf_info(model_arch)
+    if (model_arch == "Cohere2ForCausalLM"
+            and transformers.__version__ < "4.48.0"):
```
Comment (renjie0) [inline tests/models/test_initialization.py:16]: Why is this needed

Comment (DarkLight1337) [reply tests/models/test_initialization.py:16]: The model is only available in transformers v4.48 which has not been released yet.

Comment (renjie0) [reply tests/models/test_initialization.py:16]: Thanks. That makes sense.

**Comment by ExtReMLapin** [comment]
To use it at full context size, what are you guys doing ? is ` --max-model-len` enough ?

Because model config says 

`"max_position_embeddings": 8192,`
